[XenTag] Suggest threads

maitandat1507 opened the pull request:
Khi vì một lý do nào đó người xem vào thread mà không được xem (có thể do không có quyền, thread đã bị soft delete, etc.) thì hiển thị list 20 thread khác có cùng tags (20 là option, có thể thay đổi)

maitandat1507 committed:
[Suggest Thread]
#version 1.0

daohoangson commented on library/Tinhte/XenTag/Route/Prefix/Content.php:
```diff
@@ -0,0 +1,9 @@
+<?php
+class Tinhte_XenTag_Route_Prefix_Content implements XenForo_Route_Interface
```
Cái route này có vẻ không xài?

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,67 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+        try {
+            return parent::actionIndex();
+        } catch (XenForo_ControllerResponse_Exception $e) {
+            XenForo_Phrase::loadPhrases('tinhte_xentag_sorry_this_thread_can_not_display');
```
loadPhrases để làm gì vậy?

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,67 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+                $threadModOptions = $threadModel->addInlineModOptionToThread($thread, $forum);
+                $inlineModOptions += $threadModOptions;
+
+                $thread = $threadModel->prepareThread($thread, $forum);
```
Đoạn này sai. Tham số thứ 2 cho `prepareThread` phải là forum của cái thread đó. Hiện đang lấy forum có id là `$forumId[0]`, không hợp lý.

daohoangson replied:
Cái này quan trọng mà chưa fix à?

maitandat1507 replied:
Ủa ngộ vậy ta, hồi trưa em fix lên hết rồi mà ạ @@

daohoangson replied:
Cái `$forum` phải lấy cho từng thread trong mảng `$threads` chứ đâu phải lấy của error thread đâu chú.

maitandat1507 committed:
# Delete Route Prefix "suggest_thread" in AdminCP
# Delete folder "Route"
# Edit some code in XenTag_XenForo_ControllerPublic_Thread.php

maitandat1507 committed:
# Edit file "XenTag/XenForo/ControllerPublic/Thread.php"

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,66 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+            $visitor = XenForo_Visitor::getInstance();
+
+            $threadId = $this->_input->filterSingle('thread_id', XenForo_Input::UINT);
+            $numberOfContent = XenForo_Application::get('options')->Tinhte_XenTag_suggestThreads;
```
Thêm kiểm tra nếu `$numberOfContent` nhỏ hơn hoặc bằng 0 thì không chạy tiếp mà `throw $e` để người khác/add-on khác xử lý.

daohoangson replied:
Add-on này có cái hàm `Tinhte_XenTag_Option::get()` nên xài thay vì gọi `XenForo_Application::get('options')` trực tiếp.

daohoangson commented on library/Tinhte/XenTag/Model/Content.php:
```diff
@@ -0,0 +1,20 @@
+class Tinhte_XenTag_Model_Content extends XenForo_Model
+{
+    //get list content_id have the same tag with error thread
+    public function getListContentHaveTheSameTag($threadId, $numberOfContent)
```
Hàm này chắc nên cho vào `Tinhte_XenTag_Model_Search` chứ không nên tách riêng model nữa, có mỗi một hàm. Với cả tên hàm chưa rõ ràng, nên đặt là `getThreadIdsRelatedToThreadId`, cái param thứ 2 thì nên để là `$limit` cho ngắn gọn.

daohoangson commented on library/Tinhte/XenTag/Model/Content.php:
```diff
@@ -0,0 +1,20 @@
+            WHERE tag_id IN (
+                                SELECT C.tag_id
+                                FROM xf_tag_content C
+                                WHERE C.content_id = '. $threadId .'
```
Điều kiện WHERE chưa đầy đủ, cần thêm `content_type=thread` kết hợp với `content_id=$threadId` thì mới chính xác (do có thể có content type khác mà cùng id với `$threadId`).

maitandat1507 replied:
Mấy comment khác em đã fix hết rồi, chỉ còn mỗi cái chỗ này, content_type định dạng kiểu varbinary, làm sao để convert nó thành string để mình so sánh ạ?

daohoangson replied:
`VARBINARY` vẫn so sánh `= "xxx"` bình thường mà chú.

maitandat1507 replied:
À, trong câu truy vấn em để dấu ngoặc đơn nên nó không hiểu, đổi lại ngoặc kép mới được @@

daohoangson commented on library/Tinhte/XenTag/Model/Content.php:
```diff
@@ -0,0 +1,20 @@
+                                WHERE C.content_id = '. $threadId .'
+                            ) AND content_id <> '. $threadId .'
+            ORDER BY content_id DESC
+            LIMIT '. $numberOfContent .'
```
Nên xài `XenForo_Model::limitQueryResults` thay vì viết LIMIT trực tiếp trong query.

daohoangson commented on library/Tinhte/XenTag/Model/Content.php:
```diff
@@ -0,0 +1,20 @@
+    //get list content_id have the same tag with error thread
+    public function getListContentHaveTheSameTag($threadId, $numberOfContent)
+    {
+        return $this->fetchAllKeyed('
```
Chỗ này chắc không cần `fetchAllKeyed`, sử dụng `_getDb()->fetchCol()` có vẻ hợp lý hơn.

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,66 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+        }
+    }
+
+    protected function _getForumFetchOptions()
```
Hàm này không xài, xóa đi.

daohoangson commented on library/Tinhte/XenTag/addon-Tinhte_XenTag.xml:
```diff
@@ -509,6 +509,14 @@ disabled = {xen:phrase tinhte_xentag_auto_tag_mode_disabled}</edit_format_params
+    <option option_id="Tinhte_XenTag_suggestThreads" edit_format="spinbox" data_type="integer" can_backup="1">
+      <default_value>20</default_value>
+      <edit_format_params>Min = 0
+Max = 99999
```
Sao lại set max = 99999?

maitandat1507 replied:
Do em chưa biết nên để bao nhiêu nên cho nó nhiều nhiều tí luôn ạ, vậy mình set bao nhiêu là được anh?

daohoangson replied:
Không set max cũng được mà.

maitandat1507 committed:
[XenTag] Suggest threads
# Delete Model/Content.php
# Add a method to "Model/Search.php" to get threadIds related to threadId
# Edit Option, delete Max definition.

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,59 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+            $threadId = $this->_input->filterSingle('thread_id', XenForo_Input::UINT);
+            $numberOfContent = Tinhte_XenTag_Option::get('suggestThreads');
+            if ($numberOfContent <= 0) {
+                throw new XenForo_Exception($e);
```
`throw $e` luôn, không tạo exception mới nữa.

daohoangson commented on library/Tinhte/XenTag/XenForo/ControllerPublic/Thread.php:
```diff
@@ -17,4 +17,59 @@ public function Tinhte_XenTag_actionTags(XenForo_DataWriter_Discussion_Thread $t
+            }
+
+            /** @var Tinhte_XenTag_Model_Content $contentModel */
+            $contentModel = $this->getModelFromCache('Tinhte_XenTag_Model_Search');
```
Tên biến sai, tạo model Search mà tên lại là content?!
Thêm nữa là đoạn comment `@var` ở trên cũng chưa đúng.

daohoangson commented on library/Tinhte/XenTag/Model/Search.php:
```diff
@@ -71,4 +71,23 @@ public function sortResults(array $prioritizedContents, array &$results, array $
+                '
+                    SELECT DISTINCT content_id
+                    FROM xf_tag_content
+                    WHERE tag_id IN (
```
Nếu tìm theo tag_id thôi thì chưa đúng, vì có thể trả lại các content khác. Nếu chỉ lấy thread thì query ở ngoài này cũng cần thêm `content_type = thread` mới đúng.

maitandat1507 committed:
[XenTag] Suggest threads
# Edit some code
